pkg/audit/manager.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return err
+	}
+	am.mgr = mgr
+	go startMgr(mgr, am.stopper, am.stopped)
+	go am.auditManagerLoop(ctx)
+	return nil
+}
+
+// audit audits resources periodically
+func (am *auditManager) audit() error {
+	c := am.mgr.GetClient()
+
+	// don't audit anything until the constraintTemplate crd is in the cluster
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+	if err := c.Get(context.TODO(), types.NamespacedName{Name: crdName}, crd); err != nil {
+		return err
+	}
+	resp, err := am.opa.Audit(context.TODO())
+	if err != nil {
+		return err
+	}
+	log.Info("Audit opa.Audit() audit results", "violations", len(resp.Results()))
+	dynamic := dynamic.NewForConfigOrDie(am.cfg)
+
+	if len(resp.Results()) > 0 {
+		updateList := make(map[string][]auditResult)
+
+		for _, r := range resp.Results() {
+			name := r.Constraint.GetName()
+			apiVersion := r.Constraint.GetAPIVersion()
+			gvk := r.Constraint.GroupVersionKind()
+			kind := gvk.Kind
+			group := gvk.Group
+			version := gvk.Version
+			selfLink := r.Constraint.GetSelfLink()
+			message := r.Msg
+			if len(message) > msgSize {
+				message = truncateString(message, msgSize)
+			}
+			resource, ok := r.Resource.(*unstructured.Unstructured)
+			if !ok {
+				return errors.Errorf("could not cast resource as reviewResource: %v", r.Resource)
+			}
+			rname := resource.GetName()
+			rkind := resource.GetKind()
+			rnamespace := resource.GetNamespace()
+
+			updateList[selfLink] = append(updateList[selfLink], auditResult{
+				cgroup:      group,
+				cversion:    version,
+				capiversion: apiVersion,
+				ckind:       kind,
+				cname:       name,
+				rkind:       rkind,
+				rname:       rname,
+				rnamespace:  rnamespace,
+				message:     message,
+			})
+		}
+		// get all constraints kind
+		clientset := kubernetes.NewForConfigOrDie(am.cfg)
+		rs, err := clientset.Discovery().ServerResourcesForGroupVersion(constraintsGV)
+		if err != nil {
+			return err
+		}
+		// get constraints for each kind
+		for _, r := range rs.APIResources {
+			log.Info("constraint", "resource kind", r.Kind)
+			resourceGvk := strings.Split(rs.GroupVersion, "/")
+			group := resourceGvk[0]
+			version := resourceGvk[1]
+			resourceClient := dynamic.Resource(schema.GroupVersionResource{Group: group, Version: version, Resource: r.Name})
+			l, err := resourceClient.List(metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+			// get and clear/reset each constraint violation if it's not currently empty
+			for _, item := range l.Items {
+				oname := item.GetName()
+				oapiVersion := item.GetAPIVersion()
+				ogvk := item.GroupVersionKind()
+				okind := ogvk.Kind
+				ogroup := ogvk.Group
+				oversion := ogvk.Version
+				violations, found, err := unstructured.NestedSlice(item.Object, "status", "violations")
+				if err != nil {
+					return err
+				}
+				if !found || len(violations) == 0 {
+					fmt.Printf("constraint %s violations is either not found or empty, skip clear \n", oname)
```

Comment:
Let's get rid of these printfs before we merge the PR and/or replace them with logging calls if we think they're useful for the administrator.